### ISSUE TITLE
Question on hard example mining

### ISSUE THREAD

**Issue by nicedoctor** [Li-Wanhua/POEs#6]
In your paper, I have noticed that the hinge loss resquests that |yl-ym| < |yl-yn|, however , you choose the triplet by minimize | |yl-ym| - |yl -yn| | in the supplementary material.  It can only ensure the  the difference between |yl − ym| and |yl − yn| is small,  but can't  promise that |yl- ym| is smaller than |yl - yn|, just like you did in your code:

target_dis = torch.abs(target_dis - torch.abs(
            target[anchor_pos] - target[second_pos]).view(-1, 1).repeat(1, batch_size))
offset_m = torch.eye(batch_size).cuda().to(dtype=target_dis.dtype)
target_dis = target_dis + offset_m * 1000
target_dis[target_dis == 0] = 700
thrid_pos = torch.argmin(target_dis, dim=1)

I think the first line can be changed as follows:

target_dis = torch.clamp((target_dis - torch.abs(
           target[anchor_pos] - target[second_pos]).view(-1, 1).repeat(1, batch_size)),0)

use torch.clamp to ensure |yl-ym| < |yl-yn|,
and the negetive value(have changed to 0) will be 700 after the orign code:  target_dis[target_dis == 0] = 700

**Comment by nicedoctor** [comment]
However, I noticed that you use a torch.sign() to make sure that the smaller difference pair to act as the real |yl - ym|
